src/compiler/crystal/tools/doc/generator.cr

```diff
@@ -251,7 +243,22 @@ class Crystal::Doc::Generator
     end
   end
 
-  # adds extra line break to TODO lines
+  # Replaces flag keywords with html equivalent
+  # 
+  # Assumes that flag keywords are at the beginning of respective `p` element
+  def generate_flags(string)
+    string.split("<p>").map do |text|
+      text
+        .gsub(/^ ?TODO(:| )?/, %(<span class="flag orange">TODO</span> ))
+        .gsub(/^ ?FIXME(:| )?/, %(<span class="flag red">FIXME</span> ))
+        .gsub(/^ ?BUG(:| )?/, %(<span class="flag red">BUG</span> ))
+        .gsub(/^ ?NOTE(:| )?/, %(<span class="flag yellow">NOTE</span> ))
+        .gsub(/^ ?OPTIMIZE(:| )?/, %(<span class="flag green">OPTIMIZE</span> ))
+        .gsub(/^ ?DEPRECATED(:| )?/, %(<span class="flag red">DEPRECATED</span> ))
+    end.join("<p>")
+  end
+
+  # Adds extra line break to flag keyword lines
   def isolate_flag_lines(string)
     string.split(/\n/).map do |line|
       if line =~ /^ ?(TODO|FIXME|BUG|NOTE|OPTIMIZE|DEPRECATED):?/
```

Comment:
I'd use a hash containing flags and corresponding colors — or even just an array with flags, moving the colors into the css itself.